src/sinks/loki/sink.rs

```diff
@@ -163,35 +163,100 @@ pub(super) struct EventEncoder {
+                    warn!(
+                        "Failed to expand dynamic label. value: {}, err: {}",
+                        value_s,
+                        output.err().unwrap()
+                    );
+                    continue;
+                }
+
+                // key_* -> key_one, key_two, key_three
+                // * -> one, two, three
+                for (k, v) in output.unwrap() {
+                    let key = slugify_text(format!("{}{}", opening_prefix, k));
+                    let val = Value::from(v).to_string_lossy().into_owned();
+                    if val == "<null>" {
+                        warn!("Encountered \"null\" value for dynamic label. key: {}", key);
                     }
-                } else {
-                    static_vec.insert(key, value);
+                    if let Some(prev) = dynamic_labels.insert(key.clone(), val.clone()) {
+                        warn!(
+                            "Encountered duplicated dynamic label. \
+                                key: {}, value: {}, discarded value: {}",
+                            key, val, prev
+                        );
+                    };
                 }
+            } else {
+                static_labels.insert(key_s, value_s);
             }
         }
-        dynamic_vec.extend(static_vec);
-        Vec::from_iter(dynamic_vec)
+
+        for (k, v) in static_labels {
+            if let Some(discarded_v) = dynamic_labels.insert(k.clone(), v.clone()) {
+                warn!(
+                    "Static label overrides dynamic label. \
+                key: {}, value: {}, discarded value: {}",
+                    k, v, discarded_v
+                );
+            };
+        }
+
+        Vec::from_iter(dynamic_labels)
+    }
+
+    fn warn_on_label_expansion_error(
+        key_template: &Template,
+        value_template: &Template,
+        key: &Result<String, crate::template::TemplateRenderingError>,
+        value: &Result<String, crate::template::TemplateRenderingError>,
+    ) {
+        if key.is_err() && value.is_err() {
+            warn!(
+                "Failed to expand label key and value: \
+                key_template: {}, key_err: {}, value_template: {}, value_err: {}",
+                key_template,
+                key.as_ref().err().unwrap(),
+                value_template,
+                value.as_ref().err().unwrap()
+            );
+        } else if key.is_err() {
+            warn!(
+                "Failed to expand label key: key_template: {}, key_err: {}, value: {}",
+                key_template,
+                key.as_ref().err().unwrap(),
+                value.as_ref().unwrap()
+            );
+        } else {
+            warn!(
+                "Failed to expand label value: key: {}, value_template: {}, value_err: {}",
+                key.as_ref().unwrap(),
+                value_template,
+                value.as_ref().err().unwrap()
+            );
+        }
```

Comment:
This feels a little repetitive to me, I'm not 100% on this - but, what do you think about using an `internal_event` here and emitting two if both the key and value errored?

Reply:
Hi @spencergilbert,

thank you for the review and feedback.

Yes, this is repetitive and I was not really sure if this is the best approach, or if two warnings would be better. I'll simplify that as suggested.

Reply:
Switch to `emit` and one event for key and one event for value errors.
In-lined function as is shorter this way.